cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "scan.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/scan.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace detail {
+
+template <typename T>
+using pair_type = thrust::pair<T, T>;
+
+/**
+ * @brief functor to be summed over in a prefix sum such that
+ * the recurrence in question is solved.
+ * see https://www.cs.cmu.edu/~guyb/papers/Ble93.pdf S. 1.4
+ * for details
+ */
+template <typename T>
+class recurrence_functor {
+ public:
+  __device__ pair_type<T> operator()(pair_type<T> ci, pair_type<T> cj)
+  {
+    return {ci.first * cj.first, ci.second * cj.first + cj.second};
+  }
+};
+
+template <typename T>
+class ewma_functor_base {
+ public:
+  pair_type<T> IDENTITY = {1.0, 0.0};
+};
+
+template <typename T, bool nulls, bool is_numerator>
+class ewma_adjust_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_adjust_functor(T beta) : beta{beta} {}
+
+  using tupletype = std::conditional_t<nulls, thrust::tuple<bool, int, T>, T>;
+
+  __device__ pair_type<T> operator()(tupletype data)
+  {
+    if constexpr (is_numerator) {
+      if constexpr (nulls) {
```

Comment:
Try swapping the order of the two conditionals. From looking at the code, the nullable code paths are nearly identical whether we're calculating the numerator or the denominator. You should be able to make a much smaller `if constexpr (is_numerator)` block inside an `if constexpr (nulls)` block.